File: src/cpp/testAlgorithms.cpp
#include "algorithm.h"
#include <cassert>
#include <iostream>
#include <queue>

void testFibonacci() {
  assert(fib(0) == 0);
  assert(fib(1) == 1);
  assert(fib(2) == 1);
  assert(fib(3) == 2);
  assert(fib(4) == 3);
  assert(fib(5) == 5);
  assert(fib(10) == 55);
  std::cout << "Tests fibonacci OK" << std::endl;
}

void testLinearSearch() {
  int array[] = {1, 2, 3, 4, 5, 6, 7, 8};
  assert(linearSearch(array, 8, 7) == 6);
  assert(linearSearch(array, 8, 33) == -1);
  std::cout << "Tests linearSearch OK" << std::endl;
}

void testBinarySearch() {
  int sortedArray[] = {1, 2, 3, 4, 5, 6};
  assert(binarySearch(sortedArray, 6, 5) == 4);
  assert(binarySearch(sortedArray, 6, 33) == -1);
  std::cout << "Tests binarySearch OK" << std::endl;
}

bool cmpArrays(int x[], int y[], int length) {
  for (int i = 0; i < length; i++) {
    if (x[i] != y[i])
      return false;
  }
  return true;
}

void testBubbleSort() {
  int unsortedArray[] = {3, 9, 10, 8, 7, 5, 2, 6, 1, 4};
  int sortedArray[] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
  int anotherSortedArray[] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
  bubbleSort(unsortedArray, 10);
  assert(cmpArrays(unsortedArray, sortedArray, 10));
  bubbleSort(anotherSortedArray, 10);
  assert(cmpArrays(anotherSortedArray, sortedArray, 10));
  std::cout << "Tests bubbleSort OK" << std::endl;
}

void testQueue() {
  Queue queue = Queue();
  assert(queue.isEmpty());
  assert(queue.push(10) == 0);
  assert(!queue.isEmpty());
  assert(queue.push(11) == 1);
  assert(queue.peek() == 10);
  assert(queue.pop() == 10);
  assert(queue.peek() == 11);
  assert(queue.pop() == 11);
  assert(queue.isEmpty());
  std::cout << "Tests Queue OK" << std::endl;
}

int main() {
  testFibonacci();
  testLinearSearch();
  testBinarySearch();
  testBubbleSort();
  testQueue();
  std::cout << "Todos los tests han pasado" << std::endl;
  return 0;
}

File: src/js/testAlgorithms.js
const { fibonacci, linearSearch, binarySearch, bubbleSort, Queue } = require(
  "./algorithm.js",
);
const assert = require("assert");

function testFibonacci() {
  assert(fibonacci(0) == 0);
  assert(fibonacci(1) == 1);
  assert(fibonacci(2) == 1);
  assert(fibonacci(3) == 2);
  assert(fibonacci(4) == 3);
  assert(fibonacci(5) == 5);
  assert(fibonacci(10) == 55);
  console.log("Fibonacci tests passed");
}

function testLinearSearch() {
  const array = new Int32Array([1, 2, 3, 4, 5, 6, 7, 8]);
  assert(linearSearch(array, 7) == 6);
  assert(linearSearch(array, 33) == -1);
  console.log("LinearSearch tests passed");
}

function testBinarySearch() {
  const sortedArray = new Int32Array([1, 2, 3, 4, 5, 6, 7, 8]);
  assert(binarySearch(sortedArray, 7) == 6);
  assert(binarySearch(sortedArray, 33) == -1);
  console.log("BinarySearch tests passed");
}

function cmpArrays(x, y) {
  if (x.length !== y.length) {
    return false;
  } else {
    for (let i = 0; i < x.length; i++) {
      if (x[i] !== y[i]) {
        return false;
      }
    }
    return true;
  }
}

function testBubbleSort() {
  const unsortedArray = [3, 9, 10, 8, 7, 5, 2, 6, 1, 4];
  const sortedArray = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
  const anotherSortedArray = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
  bubbleSort(unsortedArray);
  assert(
    cmpArrays(unsortedArray, sortedArray),
    "test bubbleSort(unsortedArray)",
  );
  bubbleSort(anotherSortedArray);
  assert(
    cmpArrays(anotherSortedArray, sortedArray),
    "test bubbleSort(anotherSortedArray)",
  );
  console.log("BubbleSort tests passed");
}

function testQueue() {
  const queue = new Queue();
  assert(queue.isEmpty(), "queue is empty");
  assert(queue.push(8) === 0, "push(8)");
  assert(!queue.isEmpty(), "queue is not empty");
  assert(queue.push(6) === 1, "push(6)");
  assert(queue.peek() === 8, "peek()===8");
  assert(queue.push(5) === 2, "push(5)");
  assert(cmpArrays(queue.items, [8, 6, 5]), "queue elements");
  assert(queue.pop() === 8, "pop 8");
  assert(queue.peek() === 6, "peek()===6");
  assert(cmpArrays(queue.items, [6, 5]), "pop 8 cmp");
  assert(queue.pop() === 6, "pop 6");
  assert(!queue.isEmpty(), "queue not is empty");
  assert(queue.pop() === 5, "pop 5");
  assert(queue.isEmpty(), "queue is empty");
  assert(cmpArrays(queue.items, []), "queue elements");
  console.log("Queue tests passed");
}

function main() {
  testFibonacci();
  testLinearSearch();
  testBinarySearch();
  testBubbleSort();
  testQueue();
  console.log("Tests passed- everything looks OK!");
}

main();
